packages/xo-server/src/index.js

```diff
@@ -140,18 +148,38 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  let currentUser
   const SIGNIN_STRATEGY_RE = /^\/signin\/([^/]+)(\/callback)?(:?\?.*)?$/
   express.use(async (req, res, next) => {
-    const { url } = req
+    const { url, body } = req
     const matches = url.match(SIGNIN_STRATEGY_RE)
 
     if (matches) {
       return passport.authenticate(matches[1], async (err, user, info) => {
+        if (user) {
+          currentUser = user
+        }
+
         if (err) {
           return next(err)
         }
 
-        if (!user) {
+        // OTP authent
+        const otpSecret = currentUser?.preferences?.otp
+        if (otpSecret !== undefined) {
+          const isValid = authenticator.check(body.otp, otpSecret)
+
+          if (body.otp !== undefined && !isValid) {
+            req.flash('otp', true)
+            req.flash('error', 'Invalid code')
+            return res.redirect('/signin')
+          } else if (!isValid) {
+            req.flash('otp', true)
+            return res.redirect('/signin')
+          }
+        }
+
+        if (!currentUser) {
```

Comment:
I feel like the flow of this function is a bit complicated, can we simplify it?

Reply:
Do you have any idea? Because when we load the "token" page, we lose the user information

Reply:
Maybe add the user in a flash?

Reply:
I thought about it, but I didn't want to pass user information through it (security ?). But if you validate, then I will.

Reply:
AFAIU, this info is stored server side in a session and not transmitted to the user, you can store only the id to limit issues on this.